File: 4242-sum-of-gcd-of-formed-pairs/sum-of-gcd-of-formed-pairs.cpp
class Solution {
public:
    long long gcdSum(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefixGcd(n);
        int mx = 0;

        for(int i = 0; i < n; i++) {
            mx = max(mx, nums[i]);
            prefixGcd[i] = gcd(nums[i], mx);
        }

        sort(prefixGcd.begin(), prefixGcd.end());

        int i = 0;
        int j = n-1;
        long long ans = 0;

        while(i < j) {
            ans += gcd(prefixGcd[i], prefixGcd[j]);
            i++;
            j--;
        }

        return ans;
    }

    int gcd(int a, int b) {
        while(b > 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
};

File: 4242-sum-of-gcd-of-formed-pairs/sum-of-gcd-of-formed-pairs.java
class Solution {
    public long gcdSum(int[] nums) {
        int n = nums.length;
        int max = 0;
        int[] prefixGcd = new int[n];

        for(int i = 0; i < n; i++) {
            max = Math.max(max, nums[i]);
            prefixGcd[i] = GCD(nums[i], max);
        }

        Arrays.sort(prefixGcd);

        long ans = 0;
        int i = 0;
        int j = n - 1;

        while(i < j) {
            ans += GCD(prefixGcd[i], prefixGcd[j]);
            i++;
            j--;
        }

        return ans;
    }
    private int GCD(int a, int b) {
        return b == 0 ? a : GCD(b, a % b);
    }
}